### PR TITLE
Extend Unsigned Integers to Carry Units of Measure

### PR DESCRIPTION
**Pull request by pblasucci**
So, this work is in response to this suggestion: https://github.com/fsharp/fslang-suggestions/issues/901. ~~But, there's no associated RFC. If one is needed, I'd be happy to pull it together. However, at the point, the changes seem complete.~~

Also, this has only been tested on Linux. And it's my first patch to the compiler. So if there are any issues (I expect two QA tests to fail), or any changes needed, just let me know. 

Thanks!

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=9978) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=9978) before we can accept your contribution.<br/><br/>:x: pblasucci [sign now](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=9978)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/fsharp?pullRequest=9978) it.</sub>

**Comment by cartermp** [comment]
Thanks, good work so far. Looks like there are two fsharpqa failures:

`E_UnsupportedTypes01`

and

`E_UnsupportedType01`

**Comment by pblasucci** [comment]
Thanks, @cartermp! 

I was actually expecting those two failures. And, now that I've seen them (thus validating my understanding of things), I've submitted a patch. It also updates the error message around supported types.

**Comment by Happypig375** [comment]
No love for System.Numerics.Complex?

**Comment by pblasucci** [comment]
@Happypig375 There will likely be separate work to deal with the "non-primitive" numeric types (`BigInt`, `Complex`, and so on). But they require a very different mechanism than what's used for the "built-in" primitives. So, baby steps.  :wink:

**Comment by abelbraaksma** [comment]
Relevant discussion thread for the underlying RFC is here: https://github.com/fsharp/fslang-design/issues/496, though the scope here is deliberately only as in the language suggestion: primitive types.

**Comment by pblasucci** [comment]
trying to re-trigger the build

**Comment by pblasucci** [comment]
Asking these here (as opposed to Slack), in the hopes that it reaches a wider audience (and, yes, it's relevant to the currently failing builds):

+ Is there a reason why "decimals" are handled _differently_ from the other primitive numerics in `TcGlobals.fs`?

Specifically, the `TType` for decimals is loaded _directly_ from the system assemblies; while for the other, the `TType` is derived from the types (non-local?) `EntityRef`. 

And, somewhat related:

+  Is there clear guidance for when to hand the type checker a `TType` versus an `EntityRef`?

**Comment by cartermp** [comment]
> Is there clear guidance for when to hand the type checker a TType versus an EntityRef?

A TType is a more abstract type that can be transformed into an EntityRef (which has rich information about what is being inspected). I don't think there's much clean guidance here other than `TType` being used in most places because we could be handling multiple kinds of TTypes in a given expression (e.g., pretty-printing). So if it feels like a more generalized case then a TType is better, but if you're always working EntityRef data then that would be good to pass instead, i suppose.

**Comment by pblasucci** [comment]
Re-trigger the build

**Comment by abelbraaksma** [comment]
@cartermp, I think that the `make rfc` label can be removed, there's an RFC meanwhile.

**Comment by cartermp** [comment]
@pblasucci you'll want to also guard changes with langversion checks. See [here](https://github.com/dotnet/fsharp/pull/9400/files?file-filters%5B%5D=.bsl&file-filters%5B%5D=.fs&file-filters%5B%5D=.txt#diff-af6007718e25131307f300d76102d6ac) for an example of what to add and how to do the checks.

File: src/fsharp/FSharp.Core/prim-types.fsi
Comment (cartermp) [inline src/fsharp/FSharp.Core/prim-types.fsi:990]: These will need `[<Experimental("Experimental library feature, requires '--langversion:preview'")>]` for each additional API

File: tests/FSharp.Core.UnitTests/FSharp.Core/PrimTypes.fs
Comment (cartermp) [inline tests/FSharp.Core.UnitTests/FSharp.Core/PrimTypes.fs:48]: These would need to be guarded against preview somehow, @KevinRansom how do we do the testing for that with FSharp.Core APIs?

File: src/fsharp/FSharp.Core/prim-types.fsi
Comment (pblasucci) [reply src/fsharp/FSharp.Core/prim-types.fsi:991]: Done!

File: tests/FSharp.Core.UnitTests/FSharp.Core/PrimTypes.fs
Comment (pblasucci) [reply tests/FSharp.Core.UnitTests/FSharp.Core/PrimTypes.fs:48]: I _think_ I've managed this already. But it was so easy that I'm a tad skeptical. :upside_down_face:

**Comment by pblasucci** [comment]
So, I'm fairly certain all the guards and tests and such are in place for this now. In other words, unless someone has some feedback/changes, I _think_ it's ready to be moved to the next phase. Thoughts?

**Comment by cartermp** [comment]
Yeah, I think this is generally good. The tricky bit now is that we're in lockdown mode for F# 5 we may want to hold off on any additions just for the sake of stability. Since this would come in under a preview flag it wouldn't be a part of F# 5, so it's less about introducing something late in the game as it is introducing some churn that could potentially destabilize us.

I'll defer to @KevinRansom on if we want to take this now or later.

**Comment by abelbraaksma** [comment]
This looks great! Fwiw, it looks like relatively low-risk to still add to the F# 5.0 branch and considering the long-standing wish to have this in for all primitive types, and the cool bug-fix to support all build-in primitive aliases makes one wish this makes the threshold :100: :pray:.

**Comment by cartermp** [comment]
Update: we're closed down for F# 5 (barring enormous bugs) and we'll have a plannign session soon for the next sprint. I intend to bringing this to the table.

File: src/fsharp/TypeChecker.fs
Comment (KevinRansom) [inline src/fsharp/TypeChecker.fs:932]: please rename to something more descriptive ... E.g. includeExpandedMeasurables or perhapsexpandedMeasurablesEnabled

File: tests/FSharp.Core.UnitTests/FSharp.Core/PrimTypes.fs
Comment (KevinRansom) [reply tests/FSharp.Core.UnitTests/FSharp.Core/PrimTypes.fs:48]: FSharp.Core unit tests are compiled to test as LangeVersion Preview.  So this is fine.

**Review by KevinRansom — APPROVED** [review]
Small naming change, otherwise this is great.

**Comment by KevinRansom** [comment]
@pblasucci Thanks for this it is a great addition to the language.

File: src/fsharp/TypeChecker.fs
Comment (pblasucci) [reply src/fsharp/TypeChecker.fs:932]: Makes good sense. I'm on it. :+1:

**Comment by KevinRansom** [comment]
bugger should have merged it before don got happy chopping up the TypeChecker.  I will fix the conflicts and merge tomorrow.  This will likely be preview for an age.

**Comment by KevinRansom** [comment]
@pblasucci, hey I don't have permission to write back to your branch.  However, a local merge on your machine, works with out conflict.  So I will let you merge it yourself :-)

**Comment by pblasucci** [comment]
@KevinRansom No worries, sir.  :-)  I've done as you suggested and we're back to green now.

**Review by cartermp — APPROVED** [review]

**Review by dsyme — APPROVED** [review]

**Comment by cartermp** [comment]
Thanks @pblasucci!

**Pull request merged**
